versioned_hdf5/subchunk_map.py

```diff
@@ -336,3 +674,861 @@ def as_subchunk_map(
+            wholes[tuple(widx_nd)] += 1
+
+    # Slice chunk_states
+    indexers = tuple(mapper.chunks_indexer for mapper in mappers)
+    indexers = _independent_ndindex(indexers)
+    states = chunk_states[indexers]
+    assert states.ndim == chunk_states.ndim == len(mappers)
+
+    mask = states != 0
+    if plus_whole:
+        wholes = wholes[indexers]
+        assert wholes.shape == mask.shape
+        mask |= wholes == len(mappers)
+
+    indices = []
+    for axis, mapper in enumerate(mappers):
+        # Get all the chunk indices involved in the selection
+        indices_i = np.asarray(mapper.chunk_indices)
+        assert indices_i.shape == (states.shape[axis],)
+        
+        # Change shape to one that matches the states
+        new_shape = tuple(-1 if i == axis else 1 for i in range(chunk_states.ndim))
+        indices_i = indices_i.reshape(new_shape)
+        indices_i = np.broadcast_to(indices_i, states.shape)
+        # Get only the indices of modified chunks
+        indices_i = indices_i[mask]
+        indices.append(indices_i.tolist())
+
+    values_idx = (states[mask] - 1).tolist()
+    return [(tuple(idx), value_idx) for *idx, value_idx in zip(*indices, values_idx)]
+
+
+def _hypercubes_between(
+    idx_from: tuple[int, ...],
+    idx_to: tuple[int, ...],
+    shape: tuple[int, ...],
+) -> Iterator[tuple[tuple[int, ...], tuple[int, ...]]]:
+    """Given two ordered points in a N-dimensional hyperspace, yield the top-left
+    corner (included) and bottom-right corner (excluded) of the N-dimensional
+    hypercubes that cover the whole space between the two points (both excluded).
+
+    In other words, fill the flattened vector in the range
+    ]product(*idx_from), product(*idx_to)[
+
+    Example:
+    --------
+    Given two points at coordinates (3, 4, 5) and (3, 7, 8)
+    in a space of shape (10, 10, 10), this function yields
+
+    - [ 3,   4, 6:] -> (3, 4, 6), (4, 5, 10)
+    - [ 3, 5:7,  :] -> (3, 5, 0), (4, 7, 10)
+    - [ 3,   7, :8] -> (3, 7, 0), (4, 8,  8)
+    """
+    ndim = len(idx_from)
+
+    if ndim == 0:
+        pass
+
+    elif ndim == 1:
+        i, j = idx_from[0], idx_to[0]
+        if i + 1 < j:
+            yield (i + 1,), (j,)
+
+    elif idx_from[0] == idx_to[0]:
+        for ni, nj in _hypercubes_between(idx_from[1:], idx_to[1:], shape[1:]):
+            yield (idx_from[0], *ni), (idx_from[0] + 1, *nj)
+
+    else:
+        for ni, nj in _hypercubes_between(
+            idx_from[1:],
+            shape[1:],
+            shape[1:],
+        ):
+            yield (idx_from[0], *ni), (idx_from[0] + 1, *nj)
+
+        if idx_from[0] + 1 < idx_to[0]:
+            yield (
+                (idx_from[0] + 1,) + (0,) * (ndim - 1),
+                (idx_to[0], *shape[1:]),
+            )
+
+        for ni, nj in _hypercubes_between(
+            (-1,) * (ndim - 1),
+            idx_to[1:],
+            shape[1:],
+        ):
+            yield (idx_to[0], *ni), (idx_to[0] + 1, *nj)
```

Comment:
Yield, tuples, etc. are all things that compile in cython but give you ZERO performance benefit from pure python.
Hacks to make it stay in C space for longer are very much possible but at the cost of some complication.